File: scripts/bindings/ggplot_pipeline_compositions.py
"""FIX-GG04: continuous limit/vector/transform compositions in actual hosts."""
import copy,json,math,sys
from pathlib import Path
ROOT=Path(__file__).resolve().parents[2]
sys.path[:0]=[str(ROOT/'packages/python'),str(Path(sys.argv[1]).resolve())]
import finstack_chart as c
out=Path(sys.argv[2]);out.mkdir(parents=True,exist_ok=True)
registry=c.ExtensionRegistry.example();records=[]
output=c.Output((ROOT/'fixtures/capability/fonts/NotoSans-Regular.ttf').read_bytes());options=c.export_options(640,360).dpi(96).basis('current')
family='continuous'
cases=json.loads((ROOT/'fixtures/parity/ggplot2/continuous-pipeline-compositions.json').read_text())['cases']
def descriptor(t):
 call={'operation':{'id':t.get('operation','example.scale_vector'),'version':'1'},'parameters':{'mode':t['mode'],'family':t['family']}}
 result={'missing_paint_is_na':t['channel']=='colour','training':'Eligible','function':{'Interpolated':{'normalization':{'Ggplot':{'family':'Linear','domain':[1,10],'reverse':False,'rescaler':'Range'}},'output':'Identity','unknown':{'kind':'Missing'}}},'ggplot':{'Continuous':{'limits':[1,10],'oob':'Censor'}} if family=='continuous' else {'Binned':{'limits':[1,10],'oob':'Squish','breaks':{'Nice':5},'right':True}},'guide':'Hidden' if t['guide_mode']=='hidden' else ({'Continuous':{'labels':'Automatic'}} if family=='continuous' else {'BinnedLegend':'Automatic'}),'oob_function':call,'rescaler_function':call,'palette_function':{'operation':{'id':'example.scale_palette','version':'1'},'parameters':{'mode':'full','channel':t['channel']}}}
 normal=result['function']['Interpolated']['normalization']['Ggplot'];normal['family']={'Log':{'base':10}} if t['transform']=='log10' else 'Linear';normal['reverse']=t['transform']=='reverse'
 result['ggplot']['Continuous']['limits']={'full':[1,10],'reversed':[10,1],'constant':[4,4]}.get(t['limit_mode'])
 if t['limit_mode'] in ('singleton','empty','null','missing'):
  result['limits_function']={'operation':{'id':'example.constant_numeric_limits','version':'1'},'parameters':{'values':{'singleton':[4],'empty':[],'null':None,'missing':[None,10]}[t['limit_mode']]}}
 return result

def layer(t):return c.points().name('marks') if t['channel']=='colour' else c.points().name('marks').numeric_scale('Size' if t['channel']=='size' else 'Alpha','v',descriptor(t))
def data_for(t,name="data"):return c.Data.columns({'x':c.column(list(map(float,range(len(t['inputs'])))),kind='float64'),'v':c.column([float(v) if v in ('Infinity','-Infinity') else v for v in t['inputs']],kind='float64').nullable(True)},name=name)
def build(t,d):
 draft=c.plot(d).with_registry(registry).profile('Ggplot2_4_0_3')
 if t['channel']=='colour':
  scale=c.color_mapped('v',descriptor(t))
  if t.get('na_mode','NA')!='NA':scale=scale.missing(paint('#00ff00' if t['na_mode']=='green' else '#00000000'))
  return draft.aes(c.aes().x('x').y(1.).color('v').color_scale('v')).scale(scale).layer(layer(t)).build()
 return draft.aes(c.aes().x('x').y(1.)).layer(layer(t)).build()
def compare(a,b):
 if isinstance(a,(int,float)) and isinstance(b,(int,float)):assert math.isclose(a,b,rel_tol=0,abs_tol=5e-14),(a,b)
 elif isinstance(a,list) and isinstance(b,list):
  assert len(a)==len(b),(a,b)
  for x,y in zip(a,b):compare(x,y)
 elif isinstance(a,dict) and isinstance(b,dict):
  assert a.keys()==b.keys(),(a,b)
  for k,v in a.items():compare(v,b[k])
 else:assert a==b,(a,b)
def paint(v):
 if v=='black':v='#000000'
 return dict(zip(('red','green','blue','alpha'),[0,0,0,0] if v is None else [int(v[i:i+2],16) for i in (1,3,5)]+[int(v[7:9],16) if len(v)==9 else 255]))
def check(t,chart,index=0):
 state=chart.semantics()['layers'][index];styles=state.get('styles',[]);wanted=t['result']['mapped']
 if t['channel']=='colour':compare([s['color'] for s in styles],[paint(v) for v in wanted if v is not None])
 if t['channel']=='alpha':compare([s['color']['alpha'] for s in styles],[0 if v in ('Infinity','-Infinity') else max(0,min(255,round((1 if v is None else v)*255))) for v in wanted])
 entries=(state.get('color_legend') or {}).get('numeric_breaks',[]) if t['channel']=='colour' else [e for group in state.get('numeric_value_guides',{}).values() for e in group]
 entries=[e for e in entries if e['visible']]
 def mapped(v):
  if v['kind'] in ('Missing','Null'):return None
  value=v['value'];return value.get('number') if isinstance(value,dict) else value
 keys=[{'values':[(-e['value'] if t['transform']=='reverse' else math.log10(e['value']) if t['transform']=='log10' else e['value']) for e in entries],'labels':[e.get('label') for e in entries],'mapped':[mapped(e['mapped']) for e in entries]}] if entries else []
 compare(keys,t['result']['keys'])
 if t['channel']=='size':compare([s['radius'].get('number') if isinstance(s['radius'],dict) else s['radius'] for s in styles],[v for v in wanted if v is not None])
 return {'styles':styles,'numeric_scales':{k:v['scale'] for k,v in state.get('numeric_scales',{}).items()},'numeric_value_guides':list(state.get('numeric_value_guides',{}).values()),'color_legend':None if state.get('color_legend') is None else {k:v for k,v in state['color_legend'].items() if k!='id'}}
for index,t in enumerate(cases):
 owned=[]
 try:
  d=data_for(t);owned.append(d);p=build(t,d);owned.append(p);wire=p.to_json();assert json.loads(wire)['version']==(39 if t['family']=='binned' and t['guide_mode']!='hidden' else 38)
  restored=c.Plot.from_json(wire,registry);owned.append(restored);assert restored.to_json()==wire
  for state in ('original','edited'):
   current=restored if state=='original' else restored.edit().layer('marks',layer(t)).build()
   if current is not restored:owned.append(current)
   chart=current.chart();owned.append(chart);actual=check(t,chart);assert 'error' not in t['result'],t
   records.append({'index':index,'state':state,**actual})
  sample=t['guide_mode']=='hidden' and t['population']=='ordinary' and ((t['limit_mode']=='full' and t['mode']=='both_index') or (t['channel']=='size' and t['transform']=='identity' and ((t['limit_mode']=='singleton' and t['mode']=='default') or (t['limit_mode'] in ('empty','missing') and t['mode']=='rescale_index'))))
  if sample:
   request=output.request(restored,options);owned.append(request);frame=request.prepare();owned.append(frame)
   for fmt in ('svg','pdf','png'):(out/f"{t['channel']}-{t['transform']}-{t['limit_mode']}-{t['mode']}.{fmt}").write_bytes(frame.export(fmt))
 except c.ChartError as error:
  assert 'error' in t['result'],(t,str(error));assert error.code in ('CHART_VALIDATION','CHART_NUMERICAL_DOMAIN'),(t,error.code)
  records.append({'index':index,'error':error.code})
 finally:
  for obj in reversed(owned):obj.dispose()
for channel in ('colour','size','alpha'):
 for limit_mode in ('full','empty'):
  selected={t['population']:t for t in cases if t['guide_mode']=='hidden' and t['channel']==channel and t['mode']=='both_index' and t['transform']=='identity' and t['limit_mode']==limit_mode};owned=[]
  try:
   original=data_for(selected['ordinary']);owned.append(original);p=build(selected['ordinary'],original);owned.append(p);chart=p.chart();owned.append(chart);wire=p.to_json();request=output.request(p,options);owned.append(request);held=request.prepare();owned.append(held);scene=held.scene()
   for population in ('missing','empty'):
    t=selected[population];replacement=data_for(t);owned.append(replacement);tx=chart.transaction();owned.append(tx);builder=tx.replace(original,replacement);owned.append(builder);update=builder.build();owned.append(update);assert 'Applied' in chart.commit(update)
    fresh=build(t,replacement);owned.append(fresh);batch=fresh.chart();owned.append(batch);actual=check(t,chart);assert actual==check(t,batch);assert p.to_json()==wire and held.scene()==scene
    records.append({'channel':channel,'limit_mode':limit_mode,'replacement':population,**actual})
  finally:
   for obj in reversed(owned):obj.dispose()
assert len(records)==3012,len(records)
(out/'records.json').write_text(json.dumps(records,sort_keys=True,allow_nan=False))
output.dispose();registry.dispose();print('PASS',len(records),'continuous composition host states')

File: crates/chart-core/tests/ggplot_pipeline_compositions.rs
//! FIX-GG04: callback limits retain their transformed order and full arity.
use chart_core::{
    ChartResult, Revision,
    grammar::*,
    interpolate::{Number, Value},
    scales::*,
};
use serde_json::{Value as Json, json};
use std::sync::{Arc, Mutex};

struct Limits;
impl CustomScaleLimits for Limits {
    fn requires_domain(&self, _: &Json) -> bool {
        false
    }
    fn descriptor(&self) -> ExtensionDescriptor {
        ExtensionDescriptor::batch("composition.limits", Revision::new(1), true)
    }
    fn validate(&self, _: &Json) -> ChartResult<()> {
        Ok(())
    }
    fn evaluate(&self, i: ScaleLimitsInput<'_>) -> ChartResult<Option<Vec<ScaleKey>>> {
        Ok(match i.parameters.as_str().unwrap() {
            "singleton" => Some(vec![ScaleKey::Number(Number(4.))]),
            "empty" => Some(vec![]),
            "null" => None,
            "missing" => Some(vec![ScaleKey::Null, ScaleKey::Number(Number(10.))]),
            _ => unreachable!(),
        })
    }
}
fn encode(v: f64) -> Json {
    if v.is_nan() {
        Json::Null
    } else if v == f64::INFINITY {
        json!("Infinity")
    } else if v == f64::NEG_INFINITY {
        json!("-Infinity")
    } else {
        json!(v)
    }
}
fn equivalent(a: &Json, b: &Json) -> bool {
    match (a, b) {
        (Json::Number(a), Json::Number(b)) => {
            (a.as_f64().unwrap() - b.as_f64().unwrap()).abs() < 5e-14
        }
        (Json::Array(a), Json::Array(b)) => {
            a.len() == b.len() && a.iter().zip(b).all(|(a, b)| equivalent(a, b))
        }
        (Json::Object(a), Json::Object(b)) => {
            a.len() == b.len()
                && a.iter()
                    .all(|(k, a)| b.get(k).is_some_and(|b| equivalent(a, b)))
        }
        _ => a == b,
    }
}
struct Vector(Arc<Mutex<Vec<Json>>>);
impl CustomScaleVector for Vector {
    fn descriptor(&self) -> ExtensionDescriptor {
        ExtensionDescriptor::batch("composition.vector", Revision::new(1), true)
    }
    fn validate(&self, _: &Json) -> ChartResult<()> {
        Ok(())
    }
    fn evaluate(&self, i: ScaleVectorInput<'_>) -> ChartResult<Option<Vec<Number>>> {
        self.0.lock().unwrap().push(json!({
            "operation": if i.stage == ScaleVectorStage::OutOfBounds { "oob" } else { "rescaler" },
            "values": i.values.iter().map(|v|encode(v.0)).collect::<Vec<_>>(),
            "range": i.limits.iter().map(|v|encode(v.0)).collect::<Vec<_>>(), "names": []
        }));
        Ok(Some(if i.stage == ScaleVectorStage::Rescale {
            if matches!(i.parameters.as_str(), Some("rescale_index" | "both_index")) {
                (1..=i.values.len())
                    .map(|v| Number(v as f64 / i.values.len() as f64))
                    .collect()
            } else {
                let a = i.limits.first().map_or(f64::NAN, |v| v.0);
                let b = i.limits.get(1).map_or(a, |v| v.0);
                if i.limits.is_empty() || (i.limits.len() == 2 && (a.is_nan() || b.is_nan())) {
                    return Err(chart_core::Diagnostic::error(
                        chart_core::DiagnosticCode::NumericalDomain,
                        "Reference callback requires comparable limits.",
                        "Use valid limits.",
                    ));
                }
                let constant = i.limits.len() == 1
                    || a == b
                    || (a != 0.
                        && b != 0.
                        && (a - b).abs() / a.abs().min(b.abs()) < 1000. * f64::EPSILON);
                i.values
                    .iter()
                    .map(|v| {
                        Number(if v.0.is_nan() {
                            f64::NAN
                        } else if constant {
                            0.5
                        } else {
                            (v.0 - a) / (b - a)
                        })
                    })
                    .collect()
            }
        } else if matches!(i.parameters.as_str(), Some("oob_index" | "both_index")) {
            (1..=i.values.len()).map(|v| Number(v as f64)).collect()
        } else {
            let a = i.limits.first().map_or(f64::NAN, |v| v.0);
            let b = i.limits.get(1).map_or(f64::NAN, |v| v.0);
            i.values
                .iter()
                .map(|v| {
                    Number(if v.0.is_finite() && (v.0 < a || v.0 > b) {
                        f64::NAN
                    } else {
                        v.0
                    })
                })
                .collect()
        }))
    }
}
struct Palette;
impl CustomScalePalette for Palette {
    fn descriptor(&self) -> ExtensionDescriptor {
        ExtensionDescriptor::batch("composition.palette", Revision::new(1), true)
    }
    fn validate(&self, _: &Json) -> ChartResult<()> {
        Ok(())
    }
    fn evaluate(&self, i: ScalePaletteInput<'_>) -> ChartResult<ScalePaletteOutput> {
        let ScalePaletteDomain::Normalized(values) = i.domain else {
            unreachable!()
        };
        Ok(ScalePaletteOutput {
            names: None,
            values: Some(
                values
                    .iter()
                    .map(|v| {
                        if v.0.is_nan() {
                            Value::Missing
                        } else {
                            match i.parameters.as_str().unwrap() {
                                "colour" => Value::Text(
                                    if v.0 < 0.5 { "#ff0000" } else { "#0000ff" }.into(),
                                ),
                                "size" => Value::number(1. + 4. * v.0),
                                _ => Value::number(v.0),
                            }
                        }
                    })
                    .collect(),
            ),
        })
    }
}
fn operation(id: &str) -> OperationRef {
    OperationRef {
        id: format!("composition.{id}"),
        version: Revision::new(1),
    }
}

fn spec(c: &Json) -> MappedScaleSpec {
    let limits = match c["limit_mode"].as_str().unwrap() {
        "automatic" => None,
        "full" => Some([Some(Number(1.)), Some(Number(10.))]),
        "reversed" => Some([Some(Number(10.)), Some(Number(1.))]),
        "constant" => Some([Some(Number(4.)), Some(Number(4.))]),
        _ => None,
    };
    let mut spec =
        MappedScaleSpec::authored(ScaleFunctionSpec::Interpolated(InterpolatedScaleSpec {
            normalization: NormalizationSpec::Ggplot {
                timestamp: None,
                family: if c["transform"] == "log10" {
                    NumericFamily::Log { base: 10. }
                } else {
                    NumericFamily::Linear
                },
                domain: [Number(1.), Number(10.)],
                reverse: c["transform"] == "reverse",
                rescaler: GgplotRescaler::Range,
            },
            output: ScaleRangeFunction::Identity,
            unknown: Value::Missing,
        }))
        .with_ggplot(GgplotScalePolicy::Continuous {
            limits,
            oob: GgplotOob::Censor,
            empty_population: false,
            nonfinite_population: false,
        })
        .unwrap()
        .with_guide(if c["guide_mode"] == "hidden" {
            GgplotScaleGuide::Hidden
        } else {
            GgplotScaleGuide::Continuous(Default::default())
        })
        .unwrap()
        .with_oob_function(ScaleVectorOperation {
            operation: operation("vector"),
            parameters: c["mode"].clone(),
        })
        .unwrap()
        .with_rescaler_function(ScaleVectorOperation {
            operation: operation("vector"),
            parameters: c["mode"].clone(),
        })
        .unwrap()
        .with_palette_function(ScalePaletteOperation {
            operation: operation("palette"),
            parameters: c["channel"].clone(),
        })
        .unwrap();
    if matches!(
        c["limit_mode"].as_str(),
        Some("singleton" | "empty" | "null" | "missing")
    ) {
        spec = spec
            .with_limits_function(ScaleLimitsOperation {
                operation: operation("limits"),
                parameters: c["limit_mode"].clone(),
            })
            .unwrap();
    }
    spec
}

#[test]
fn index_rescalers_can_ignore_empty_singleton_and_missing_limits() {
    let fixture: Json = serde_json::from_str(include_str!(
        "../../../fixtures/parity/ggplot2/continuous-pipeline-compositions.json"
    ))
    .unwrap();
    let cases = fixture["cases"]
        .as_array()
        .unwrap()
        .iter()
        .filter(|c| {
            c["population"] == "ordinary"
                && c["guide_mode"] == "hidden"
                && matches!(c["mode"].as_str(), Some("rescale_index" | "both_index"))
        })
        .collect::<Vec<_>>();
    assert_eq!(cases.len(), 144);
    let mut failures = vec![];
    for c in cases {
        let calls = Arc::new(Mutex::new(vec![]));
        let mut registry = ExtensionRegistry::new();
        registry.register_scale_limits(Arc::new(Limits)).unwrap();
        registry
            .register_scale_vector(Arc::new(Vector(calls.clone())))
            .unwrap();
        registry.register_scale_palette(Arc::new(Palette)).unwrap();
        let spec = spec(c);
        let inputs = [Some(1.), Some(4.), Some(10.)];
        let result = spec
            .trained_with_registry(&inputs.map(|v| v.map(Number)), &registry)
            .and_then(|s| MappedScale::new_with_registry(s, &registry))
            .and_then(|s| s.numeric_batch(&inputs));
        let actual = result.map(|v| {
            json!(
                v.unwrap()
                    .iter()
                    .map(|v| match v {
                        Value::Number(v) => encode(v.0),
                        Value::Text(v) => json!(v),
                        Value::Missing => Json::Null,
                        _ => unreachable!(),
                    })
                    .collect::<Vec<_>>()
            )
        });
        let expected_calls = c["calls"]
            .as_array()
            .unwrap()
            .iter()
            .filter(|v| v["operation"] != "palette")
            .cloned()
            .collect::<Vec<_>>();
        let mapped_ok = if c["result"].get("error").is_some() {
            actual.is_err()
        } else {
            actual
                .as_ref()
                .is_ok_and(|a| equivalent(a, &c["result"]["mapped"]))
        };
        if !mapped_ok || !equivalent(&json!(*calls.lock().unwrap()), &json!(expected_calls)) {
            failures.push(format!(
                "{}/{}/{}/{}: {actual:?}; calls {:?}, expected {expected_calls:?}",
                c["transform"],
                c["limit_mode"],
                c["channel"],
                c["mode"],
                calls.lock().unwrap()
            ));
        }
    }
    assert!(
        failures.is_empty(),
        "{} composition failures:\n{}",
        failures.len(),
        failures.join("\n")
    );
}

#[test]
fn all_compositions_preserve_primary_marks_and_guide_keys() {
    use chart_core::prelude::*;
    let fixture: Json = serde_json::from_str(include_str!(
        "../../../fixtures/parity/ggplot2/continuous-pipeline-compositions.json"
    ))
    .unwrap();
    let cases = fixture["cases"]
        .as_array()
        .unwrap()
        .iter()
        .collect::<Vec<_>>();
    assert_eq!(cases.len(), 1728);
    let mut failures = vec![];
    for c in cases {
        let mut registry = ExtensionRegistry::new();
        registry.register_scale_limits(Arc::new(Limits)).unwrap();
        registry
            .register_scale_vector(Arc::new(Vector(Arc::new(Mutex::new(vec![])))))
            .unwrap();
        registry.register_scale_palette(Arc::new(Palette)).unwrap();
        let inputs = c["inputs"]
            .as_array()
            .unwrap()
            .iter()
            .map(|v| match v.as_str() {
                Some("Infinity") => Some(f64::INFINITY),
                Some("-Infinity") => Some(f64::NEG_INFINITY),
                _ => v.as_f64(),
            })
            .collect::<Vec<_>>();
        let data = Data::columns()
            .column("x", (0..inputs.len()).map(|v| v as f64).collect::<Vec<_>>())
            .column("v", inputs)
            .build()
            .unwrap();
        let mut scale = spec(c);
        scale.missing_paint_is_na = c["channel"] == "colour";
        let draft = plot(data)
            .profile(Profile::Ggplot2_4_0_3)
            .extensions(Arc::new(registry));
        let plot = if c["channel"] == "colour" {
            draft
                .aes(aes().x("x").y(1.).color("v").color_scale("v"))
                .scale(color_mapped("v", scale))
                .layer(points())
                .build()
        } else {
            draft
                .aes(aes().x("x").y(1.))
                .layer(points().numeric_scale(
                    if c["channel"] == "size" {
                        NumericAesthetic::Size
                    } else {
                        NumericAesthetic::Alpha
                    },
                    "v",
                    scale,
                ))
                .build()
        }
        .unwrap();
        let result = plot.chart().unwrap().prepare();
        let matches = if c["result"].get("error").is_some() {
            result.is_err()
        } else if let Ok(prepared) = &result {
            let layer = &prepared.layers()[0];
            let entries = if c["channel"] == "colour" {
                layer
                    .color_legend()
                    .map(|v| v.numeric_breaks.clone())
                    .unwrap_or_default()
            } else {
                layer
                    .numeric_value_guides()
                    .values()
                    .flatten()
                    .cloned()
                    .collect::<Vec<_>>()
            };
            let visible = entries.iter().filter(|v| v.visible).collect::<Vec<_>>();
            let keys = if visible.is_empty() {
                json!([])
            } else {
                json!([{"values":visible.iter().map(|v|encode(if c["transform"] == "reverse" { -v.value.0 } else if c["transform"] == "log10" { v.value.0.log10() } else { v.value.0 })).collect::<Vec<_>>(),"labels":visible.iter().map(|v|v.label.clone()).collect::<Vec<_>>(),"mapped":visible.iter().map(|v|match v.mapped.as_ref().unwrap(){Value::Number(v)=>encode(v.0),Value::Text(v)=>json!(v),Value::Missing=>Json::Null,_=>unreachable!()}).collect::<Vec<_>>()}])
            };
            let wanted = c["result"]["mapped"].as_array().unwrap();
            let marks_match = if c["channel"] == "size" {
                equivalent(
                    &json!(
                        layer
                            .marks()
                            .iter()
                            .map(|m| encode(m.style.radius))
                            .collect::<Vec<_>>()
                    ),
                    &json!(wanted.iter().filter(|v| !v.is_null()).collect::<Vec<_>>()),
                )
            } else if c["channel"] == "alpha" {
                layer.marks().len() == wanted.len()
                    && layer.marks().iter().zip(wanted).all(|(m, v)| {
                        m.style.color.alpha
                            == if matches!(v.as_str(), Some("Infinity" | "-Infinity")) {
                                0
                            } else {
                                (v.as_f64().unwrap_or(1.) * 255.).round_ties_even() as u8
                            }
                    })
            } else {
                let wanted = wanted.iter().filter(|v| !v.is_null()).collect::<Vec<_>>();
                layer.marks().len() == wanted.len()
                    && layer.marks().iter().zip(wanted).all(|(m, v)| {
                        m.style.color
                            == chart_core::color::parse_r(v.as_str().unwrap())
                                .unwrap()
                                .resolve()
                    })
            };
            if !equivalent(&keys, &c["result"]["keys"]) || !marks_match {
                failures.push(format!(
                    "{}/{}/{}/{}/{}: keys {keys}, expected {}; marks_match={marks_match}",
                    c["transform"],
                    c["limit_mode"],
                    c["channel"],
                    c["population"],
                    c["guide_mode"],
                    c["result"]["keys"]
                ));
            }
            true
        } else {
            false
        };
        if !matches {
            failures.push(format!(
                "{}/{}/{}/{}/{}: actual error {:?}, expected {}",
                c["transform"],
                c["limit_mode"],
                c["channel"],
                c["population"],
                c["guide_mode"],
                result.err(),
                c["result"]
            ));
        }
    }
    assert!(
        failures.is_empty(),
        "{} primary composition failures:\n{}",
        failures.len(),
        failures.join("\n")
    );
}
